consensus/hotstuff/signature/weighted_signature_aggregator.go

```diff
@@ -1,39 +1,141 @@
-func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (totalWeight uint64, exception error) {
-	panic("to be implemented")
+// TrustedAdd adds a signature to the internal set of signatures.
+//
+// It adds the signer's weight to the total collected weight and returns the total weight regardless
+// of the returned error.
+// The function errors if a signature from the signerID was already collected.
+// The function is thread-safe
+func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (uint64, error) {
+	// get the total weight safely
+	collectedWeight := s.TotalWeight()
+
+	// get the index
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	// get the weight
+	weight, ok := s.idToWeights[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the weight map", signerID)
+	}
+
+	// atomically update the signatures pool and the total weight
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// This is a sanity check because the upper layer should have already checked for double-voters.
+	_, ok = s.collectedIDs[signerID]
+	if ok {
+		return collectedWeight, fmt.Errorf("SigneID %s was already added", signerID)
+	}
+
+	err := s.SignatureAggregatorSameMessage.TrustedAdd(index, sig)
+	if err != nil {
+		return collectedWeight, fmt.Errorf("Trusted add has failed: %w", err)
+	}
+
+	s.collectedIDs[signerID] = struct{}{}
+	collectedWeight += weight
+	s.totalWeight = collectedWeight
+	return collectedWeight, nil
 }
 
-// TotalWeight returns the total weight presented by the collected sig shares.
+// TotalWeight returns the total weight presented by the collected signatures.
+//
+// The function is thread-safe
 func (s *WeightedSignatureAggregator) TotalWeight() uint64 {
-	panic("to be implemented")
+	s.lock.RLock()
+	collectedWeight := s.totalWeight
+	s.lock.RUnlock()
+	return collectedWeight
 }
 
-// Aggregate assumes enough shares have been collected, it aggregates the signatures
-// and return the aggregated signature.
-// if called concurrently, only one thread will be running the aggregation.
-func (s *WeightedSignatureAggregator) Aggregate() ([]flow.Identifier, crypto.Signature, error) {
-	panic("to be implemented")
+// Aggregate aggregates the signatures and returns the aggregated signature.
+//
+// The function is thread-safe.
+// Aggregate attempts to aggregate the internal signatures and returns the resulting signature data.
+// The function performs a final verification and errors if the aggregated signature is not valid. This is
+// required for the function safety since "TrustedAdd" allows adding invalid signatures.
+//
+// TODO : When compacting the list of signers, update the return from []flow.Identifier
+// to a compact bit vector.
+func (s *WeightedSignatureAggregator) Aggregate() ([]flow.Identifier, []byte, error) {
```

Comment:
Only one goroutine should run the aggregation, I was imaging that we could cache aggregated signature after it's being built to avoid doing aggregation multiple times, not sure how big of an issue it would be though.

Reply:
Same [comment](https://github.com/onflow/flow-go/pull/1292#discussion_r708493745)

Reply:
And yes, I can make `SignatureAggregatorSameMessage` cache the signature.

Reply:
Implemented in https://github.com/onflow/flow-go/pull/1347